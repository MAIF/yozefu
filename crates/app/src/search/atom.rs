use async_trait::async_trait;
use lib::search::{atom::Atom, filter::Filter, offset::FromOffset};

use super::{Search, SearchContext};

#[async_trait]
impl Search for Atom {
    fn offset(&self) -> Option<FromOffset> {
        match self {
            Atom::Symbol(_) => None,
            Atom::Compare(c) => c.offset(),
            Atom::Parenthesis(c) => c.offset(),
            Atom::Filter(_) => None,
        }
    }

    fn matches(&self, context: &SearchContext) -> bool {
        match self {
            Atom::Symbol(_) => false,
            Atom::Compare(e) => e.matches(context),
            Atom::Parenthesis(e) => e.matches(context),
            Atom::Filter(f) => f.matches(context),
        }
    }

    fn filters(&self) -> Vec<Filter> {
        match self {
            Atom::Symbol(_) => vec![],
            Atom::Compare(e) => e.filters(),
            Atom::Parenthesis(e) => e.filters(),
            Atom::Filter(f) => vec![f.clone()],
        }
    }
}
