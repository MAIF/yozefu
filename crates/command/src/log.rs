//! Logging utilities.

use std::{fs::OpenOptions, path::PathBuf};
use tracing_subscriber::{
    filter::LevelFilter,
    fmt::{SubscriberBuilder, format::FmtSpan},
};

/// Returns the log level based on the debug flag.
fn log_level(is_debug: bool) -> LevelFilter {
    match is_debug {
        true => LevelFilter::DEBUG,
        false => LevelFilter::INFO,
    }
}

/// When the user starts the headless mode, it writes logs to `stderr`.
pub(crate) fn init_logging_stderr(
    is_debug: bool,
) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
    tracing_subscriber_builder(is_debug)
        //.with_writer(std::io::stderr)
        .with_ansi(true)
        .try_init()
}

/// When the user starts the TUI, it writes logs to a file.
pub(crate) fn init_logging_file(
    is_debug: bool,
    path: &PathBuf,
) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
    if let Err(e) = tracing_subscriber_builder(is_debug)
        .with_writer(OpenOptions::new().append(true).create(true).open(path)?)
        .try_init()
    {
        eprintln!("Failed to initialize logging: {}", e);
    }
    Ok(())
}

/// When the user starts the TUI, it writes logs to a file.
fn tracing_subscriber_builder(is_debug: bool) -> SubscriberBuilder {
    let level = log_level(is_debug);
    let mut builder = tracing_subscriber::fmt().with_max_level(level);
    if is_debug {
        builder = builder.with_span_events(FmtSpan::CLOSE);
    }
    builder
}
