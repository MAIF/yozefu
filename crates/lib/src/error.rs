//! Error definitions and their implementations of the `From` trait.
use rdkafka::error::KafkaError;
use std::fmt::Display;
use std::str::Utf8Error;
use std::{
    fmt::{self, Formatter},
    num::TryFromIntError,
};

/// All kinds of errors that may occur
#[derive(Debug)]
pub enum Error {
    Error(String),
    KafkaError(KafkaError),
    IoError(std::io::Error),
    SerdeError(serde_json::Error),
    ThemeError(String),
    Search(SearchError),
    SchemaRegistry(String),
    Tokio(String),
}

#[derive(Debug)]
pub enum SearchError {
    Parse(String),
}

impl std::error::Error for Error {}

impl Display for Error {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            Error::IoError(e) => write!(f, "IO Error: {}", e),
            Error::SerdeError(e) => write!(f, "Serde Error: {}", e),
            Error::ThemeError(e) => write!(f, "Theme Error: {}", e),
            Error::KafkaError(e) => write!(f, "Kafka Error: {}", e),
            Error::Error(e) => write!(f, "{}", e),
            Error::Tokio(e) => write!(f, "Tokio Error: {}", e),
            Error::Search(e) => write!(f, "{}", e),
            Error::SchemaRegistry(e) => write!(f, "Schema registry Error: {}", e),
        }
    }
}

impl Display for SearchError {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            SearchError::Parse(e) => write!(f, "Cannot parse the search query at '{}'", e),
        }
    }
}

impl From<std::fmt::Error> for Error {
    fn from(e: std::fmt::Error) -> Self {
        Error::Error(e.to_string())
    }
}

impl From<KafkaError> for Error {
    fn from(e: KafkaError) -> Self {
        Error::KafkaError(e)
    }
}

impl From<std::io::Error> for Error {
    fn from(e: std::io::Error) -> Self {
        Error::IoError(e)
    }
}

impl From<serde_json::Error> for Error {
    fn from(e: serde_json::Error) -> Self {
        Error::SerdeError(e)
    }
}

impl From<&str> for Error {
    fn from(e: &str) -> Self {
        Error::Error(e.to_string())
    }
}

impl From<TryFromIntError> for Error {
    fn from(e: TryFromIntError) -> Self {
        Error::Error(e.to_string())
    }
}

impl From<strum::ParseError> for Error {
    fn from(e: strum::ParseError) -> Self {
        Error::Error(e.to_string())
    }
}

impl From<Utf8Error> for Error {
    fn from(e: Utf8Error) -> Self {
        Error::Error(e.to_string())
    }
}
