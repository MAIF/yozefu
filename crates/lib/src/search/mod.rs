//! This module defines the the parsing functions of search query.
//! The grammar of the syntax is the following:
//!
//! ```bnf
//! search-query      ::= clause+
//! clause            ::= or-expression | limit-clause | from-clause | order-clause
//! or-expression     ::= And-expression | and-expression 'or' and-expression
//! and-expression    ::= atom | atom 'and' atom
//! term              ::= atom | '!' atom
//! atom              ::= comparison  | filter | '(' expression ')'
//! number-symbol     ::= 'offset' | 'partition' | 'size'
//! string-symbol     ::= 'topic' | 'key' | 'timestamp' | 'value'
//! symbol            ::= number-symbol | string-symbol
//! comparison        ::= number-comparison | string-comparison | time-comparison
//! number-comparison ::= number-symbol number-operator number
//! string-comparison ::= string-symbol string-operator string
//! time-comparison   ::= 'between' string 'and' string
//! number-operator   ::=  '==' | '!=' | '>' | '<' | '>=' | '<='
//! string-operator   ::= 'starts with' | '==' | '!=' | '=~' | 'contains' | 'contain' | 'includes' | 'include'
//! filter            ::= .+ '('filter-parameters')'
//! filter-parameter  ::= string | number
//! filter-parameters ::= filter-parameter  (',' filter-parameter)*
//! limit-clause      ::= 'limit' number
//! order-clause      ::= 'order by' symbol order-keyword
//! order-keyword     ::= 'asc' | 'desc'
//! from-clause       ::= 'from' offset
//! offset            ::= 'beginning' | 'begin' | 'end' | 'end' '-' number | string | number
//! number            ::= [0-9_]+
//! string            ::= '"' [^"]+ '"' | "'" [^']+ "'"
//! ```
//! You can use <https://www.bottlecaps.de/rr/ui> to visualize it.

#[cfg(feature = "native")]
pub mod atom;
#[cfg(feature = "native")]
pub mod clause;
#[cfg(feature = "native")]
pub mod expression;
#[cfg(feature = "native")]
pub mod filter;
#[cfg(feature = "native")]
pub mod number;
#[cfg(feature = "native")]
pub mod offset;
#[cfg(feature = "native")]
pub mod order;
#[cfg(feature = "native")]
pub mod search_query;
#[cfg(feature = "native")]
pub mod string;
#[cfg(feature = "native")]
pub mod symbol;
#[cfg(feature = "native")]
pub mod term;
#[cfg(feature = "native")]
pub mod timestamp;
#[cfg(feature = "native")]
pub mod wsi;

pub mod compare;

#[cfg(feature = "native")]
pub use order::Order;
#[cfg(feature = "native")]
pub use order::OrderBy;
#[cfg(feature = "native")]
pub use search_query::parse_search_query;
#[cfg(feature = "native")]
pub use search_query::SearchQuery;
use serde::Deserialize;
use serde::Serialize;

#[cfg(test)]
pub mod expression_test;
#[cfg(test)]
pub mod number_test;
#[cfg(test)]
pub mod offset_test;

#[derive(Debug, PartialEq, Clone, Default, Deserialize, Serialize)]
pub struct FilterResult {
    pub r#match: bool,
}

impl FilterResult {
    pub fn new(r#match: bool) -> Self {
        Self { r#match }
    }
}

impl From<bool> for FilterResult {
    fn from(r#match: bool) -> Self {
        Self { r#match }
    }
}
