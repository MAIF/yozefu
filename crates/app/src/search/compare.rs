use lib::{
    kafka::Comparable,
    search::{
        compare::{CompareExpression, NumberOperator, StringOperator},
        filter::Filter,
        offset::FromOffset,
    },
};

use super::SearchContext;
use crate::search::Search;

impl Search for CompareExpression {
    fn offset(&self) -> Option<FromOffset> {
        match self {
            CompareExpression::Offset(NumberOperator::Equal, e) => Some(FromOffset::Offset(*e)),
            CompareExpression::Offset(NumberOperator::GreaterOrEqual, e) => {
                Some(FromOffset::Offset(*e))
            }
            CompareExpression::Offset(NumberOperator::GreaterThan, e) => {
                Some(FromOffset::Offset(*e + 1))
            }
            CompareExpression::OffsetTail(e) => Some(FromOffset::OffsetTail(*e)),
            CompareExpression::Timestamp(op, e) => match op {
                NumberOperator::GreaterThan => Some(FromOffset::Timestamp(e.timestamp_millis())),
                NumberOperator::GreaterOrEqual => {
                    Some(FromOffset::Timestamp(e.timestamp_millis() - 1000))
                }
                NumberOperator::Equal => Some(FromOffset::Timestamp(e.timestamp_millis() - 1000)),
                NumberOperator::LowerThan => None,
                NumberOperator::LowerOrEqual => None,
                _ => None,
            },
            _ => None,
        }
    }

    fn matches(&self, context: &SearchContext) -> bool {
        let record = context.record;
        match self {
            CompareExpression::OffsetTail(_) => true,
            CompareExpression::Partition(op, p) => match op {
                &NumberOperator::GreaterThan => record.partition > *p,
                NumberOperator::GreaterOrEqual => record.partition >= *p,
                NumberOperator::LowerThan => record.partition < *p,
                NumberOperator::LowerOrEqual => record.partition <= *p,
                NumberOperator::Equal => record.partition == *p,
                NumberOperator::NotEqual => record.partition != *p,
            },
            CompareExpression::Offset(op, p) => match op {
                NumberOperator::GreaterThan => record.offset > *p,
                NumberOperator::GreaterOrEqual => record.offset >= *p,
                NumberOperator::LowerThan => record.offset < *p,
                NumberOperator::LowerOrEqual => record.offset <= *p,
                NumberOperator::Equal => record.offset == *p,
                NumberOperator::NotEqual => record.offset != *p,
            },
            CompareExpression::Topic(op, t) => match op {
                StringOperator::Equal => record.topic == *t,
                StringOperator::NotEqual => record.topic != *t,
                StringOperator::Contain => record.topic.contains(t),
                StringOperator::StartWith => record.topic.starts_with(t),
            },
            CompareExpression::Size(op, s) => match op {
                NumberOperator::GreaterThan => record.size > *s as usize,
                NumberOperator::GreaterOrEqual => record.size >= *s as usize,
                NumberOperator::LowerThan => record.size < *s as usize,
                NumberOperator::LowerOrEqual => record.size <= *s as usize,
                NumberOperator::Equal => record.size == *s as usize,
                NumberOperator::NotEqual => record.size != *s as usize,
            },
            CompareExpression::Key(op, t) => record.key.compare(&None, op, t),
            CompareExpression::Value(left, op, t) => record.value.compare(left, op, t),
            CompareExpression::Header(left, op, t) => {
                let headers = &record.headers;
                let header = headers.get(left);
                if header.is_none() {
                    return false;
                }
                let header = header.unwrap();
                match op {
                    StringOperator::Contain => header.contains(t),
                    StringOperator::Equal => header == t,
                    StringOperator::StartWith => header.starts_with(t),
                    StringOperator::NotEqual => header != t,
                }
            }
            CompareExpression::Timestamp(op, t) => {
                let ts = record.timestamp_as_local_date_time().unwrap();
                match op {
                    NumberOperator::GreaterThan => ts > *t,
                    NumberOperator::GreaterOrEqual => ts >= *t,
                    NumberOperator::LowerThan => ts < *t,
                    NumberOperator::LowerOrEqual => ts <= *t,
                    NumberOperator::Equal => ts == *t,
                    NumberOperator::NotEqual => ts != *t,
                }
            }
            CompareExpression::TimestampBetween(from, to) => {
                let ts = record.timestamp_as_local_date_time().unwrap();
                from <= &ts && &ts <= to
            }
        }
    }

    fn filters(&self) -> Vec<Filter> {
        vec![]
    }
}

#[test]
fn test_matches() {
    use crate::search::filter::CACHED_FILTERS;
    use lib::kafka::KafkaRecord;
    use std::path::PathBuf;

    let compare = CompareExpression::Offset(NumberOperator::Equal, 42);
    let record = KafkaRecord {
        topic: "test-topic".to_string(),
        partition: 0,
        offset: 42,
        key: lib::DataType::String("key".to_string()),
        value: lib::DataType::String("value".to_string()),
        timestamp: None,
        headers: std::collections::BTreeMap::new(),
        key_schema: None,
        value_schema: None,
        size: 12,
        key_as_string: "key".to_string(),
        value_as_string: "value".to_string(),
    };
    let context = SearchContext {
        record: &record,
        filters: &CACHED_FILTERS,
        filters_directory: PathBuf::from("."),
    };

    assert!(compare.matches(&context))
}
