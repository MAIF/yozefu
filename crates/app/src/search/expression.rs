use lib::search::{
    expression::{AndExpression, OrExpression},
    filter::Filter,
    offset::FromOffset,
};

use super::{Search, SearchContext};

impl Search for AndExpression {
    fn offset(&self) -> Option<FromOffset> {
        match self {
            Self::AndTerm(t) => t.offset(),
            Self::AndExpression(v) => {
                for vv in v {
                    let o = vv.offset();
                    if o.is_some() {
                        return o;
                    }
                }
                None
            }
        }
    }

    fn matches(&self, context: &SearchContext) -> bool {
        let record = context;
        match self {
            Self::AndTerm(t) => t.matches(record),
            Self::AndExpression(e) => {
                for ee in e {
                    if !ee.matches(record) {
                        return false;
                    }
                }
                true
            }
        }
    }

    fn filters(&self) -> Vec<Filter> {
        match self {
            AndExpression::AndTerm(term) => term.filters(),
            AndExpression::AndExpression(vec) => vec.iter().flat_map(|t| t.filters()).collect(),
        }
    }
}

impl Search for OrExpression {
    fn offset(&self) -> Option<FromOffset> {
        match self {
            Self::OrTerm(t) => t.offset(),
            Self::OrExpression(_) => None,
        }
    }

    fn matches(&self, context: &SearchContext) -> bool {
        match self {
            Self::OrTerm(t) => t.matches(context),
            Self::OrExpression(e) => {
                if e.is_empty() {
                    return true;
                }
                for ee in e {
                    if ee.matches(context) {
                        return true;
                    }
                }
                false
            }
        }
    }

    fn filters(&self) -> Vec<Filter> {
        match self {
            OrExpression::OrTerm(and_expression) => and_expression.filters(),
            OrExpression::OrExpression(vec) => vec.iter().flat_map(|e| e.filters()).collect(),
        }
    }
}
