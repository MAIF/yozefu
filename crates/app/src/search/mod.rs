//! Module implementing the search logic

use std::path::{Path, PathBuf};

use extism::{Manifest, Plugin, Wasm};
use filter::{CACHED_FILTERS, PARSE_PARAMETERS_FUNCTION_NAME};
use itertools::Itertools;
use lib::{
    parse_search_query,
    search::{filter::Filter, offset::FromOffset},
    KafkaRecord, SearchQuery,
};
use log::error;

pub mod atom;
pub mod compare;
pub mod expression;
pub mod filter;
pub mod search_query;
pub mod term;

pub trait Search {
    /// Returns the offset from which the search should start.
    fn offset(&self) -> Option<FromOffset> {
        None
    }
    /// returns `true` if the record matches the search query.
    fn matches(&self, context: &SearchContext) -> bool;

    /// Returns the search filters that are used in the search query.
    fn filters(&self) -> Vec<Filter>;
}

/// Struct that holds the context of the search.
/// It contains the record that is being searched and the loaded search filters.
pub struct SearchContext<'a> {
    pub record: &'a KafkaRecord,
    pub filters_directory: PathBuf,
}

impl SearchContext<'_> {
    pub fn new<'a>(record: &'a KafkaRecord, filters_directory: &'a Path) -> SearchContext<'a> {
        SearchContext {
            record,
            filters_directory: filters_directory.to_path_buf(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Default)]
pub struct ValidSearchQuery(SearchQuery);

impl ValidSearchQuery {
    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }

    pub fn limit(&self) -> Option<usize> {
        self.0.limit
    }

    pub fn query(&self) -> &SearchQuery {
        &self.0
    }
}

impl ValidSearchQuery {
    pub fn from(input: &str, filters_directory: &Path) -> Result<Self, lib::Error> {
        let query = parse_search_query(input).map_err(lib::Error::Search)?.1;
        let filters = query.filters();
        for filter in filters {
            let name = filter.name;
            let path = filters_directory.join(format!("{}.wasm", &name));
            let url = Wasm::file(&path);
            let manifest = Manifest::new([url]);
            let mut filters = CACHED_FILTERS.lock().unwrap();
            if !filters.contains_key(&name) {
                match Plugin::new(manifest, [], true) {
                    Ok(plugin) => filters.insert(name.to_string(), plugin),
                    Err(err) => {
                        error!("No such file '{}': {}", path.display(), err);
                        return Err(lib::Error::Error(format!(
                            "Cannot find search filter '{}'",
                            name
                        )));
                    }
                };
            }
            let params = filter.parameters;
            let wasm_module = &mut filters.get_mut(&name).unwrap();
            if let Err(e) = wasm_module.call::<&str, &str>(
                PARSE_PARAMETERS_FUNCTION_NAME,
                &serde_json::to_string(&params.iter().map(|e| e.json()).collect_vec()).unwrap(),
            ) {
                error!(
                    "Error when calling '{}' from wasm module '{}': {:?}",
                    PARSE_PARAMETERS_FUNCTION_NAME, name, e
                );
                return Err(lib::Error::Error(format!("{}: {e}", &name)));
            };
        }

        Ok(ValidSearchQuery(query))
    }
}

impl Search for ValidSearchQuery {
    /// Returns the offset from which the search should start.
    fn offset(&self) -> Option<FromOffset> {
        self.0.offset()
    }

    fn matches(&self, context: &SearchContext) -> bool {
        self.0.matches(context)
    }

    fn filters(&self) -> Vec<Filter> {
        self.0.filters()
    }
}
