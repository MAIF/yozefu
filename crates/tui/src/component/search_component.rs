//! This component renders the search bar.
//! It comes with the following features:
//!  - all queries are stored into a history.
//!  - The component suggests queries based on your history.

use std::{path::PathBuf, time::Duration};

use app::search::ValidSearchQuery;
use crossterm::event::{Event, KeyCode, KeyEvent, KeyModifiers};
use itertools::Itertools;
use lib::{Error, error::SearchError};
use log::error;
use ratatui::{
    Frame,
    layout::{Position, Rect},
    style::Style,
    text::{Line, Span},
    widgets::{Block, BorderType, Borders, Padding, Paragraph, Wrap},
};
use tokio::{select, sync::mpsc::UnboundedSender, time::Instant};
use tokio_util::sync::CancellationToken;
use tui_input::{Input, backend::crossterm::EventHandler};

use crate::{
    error::TuiError,
    {Action, Notification},
};

use super::{Component, ComponentName, Shortcut, State};

#[derive(Default)]
pub(crate) struct SearchComponent {
    input: Input,
    index_history: usize,
    history: Vec<String>,
    compiler_worker: CancellationToken,
    filters_directory: PathBuf,
    remaining_input: Option<String>,
    action_tx: Option<UnboundedSender<Action>>,
    autocomplete: Option<String>,
    // A hack to detect copy-paste events and replace \n with a space
    entered: Option<Instant>,
}

impl SearchComponent {
    pub fn new(input: &str, history: Vec<String>, filters_directory: PathBuf) -> Self {
        Self {
            input: Input::from(input),
            index_history: history.len().saturating_sub(1),
            history,
            filters_directory,
            ..Self::default()
        }
    }

    fn parse_input(&mut self) {
        let input = self.input.value().to_string();
        let tt = self.action_tx.clone();

        let filters_dir = self.filters_directory.clone();
        self.compiler_worker.cancel();
        self.compiler_worker = CancellationToken::new();
        let token = self.compiler_worker.clone();
        tokio::spawn(async move {
            select! {
                _ = token.cancelled() => {  },
                _ = tokio::time::sleep(Duration::from_millis(700)) => {
                    if input.len() > 5 {
                        if let Err(e) = ValidSearchQuery::from(&input, &filters_dir) {
                            error!("{e}");
                            tt.as_ref().unwrap().send(Action::Notification(Notification::new(log::Level::Error, e.to_string()))).unwrap();
                        }
                    }
                 }
            }
        });
    }

    fn autocomplete(&mut self, keycode: KeyCode) {
        let prompt = self.input.value();

        let mut possibilities = self
            .history
            .iter()
            .filter(|e| e.starts_with(prompt))
            .collect_vec();
        possibilities.dedup();

        self.index_history = match keycode {
            KeyCode::Up => match self.index_history == 0 {
                true => 0,
                false => self.index_history - 1,
            },
            KeyCode::Down => match self.index_history == possibilities.len().saturating_sub(1) {
                true => self.index_history,
                false => self.index_history + 1,
            },
            _ => match possibilities.is_empty() {
                true => 0,
                false => possibilities.len() - 1,
            },
        };
        self.autocomplete = possibilities
            .get(self.index_history)
            .map(|e| e.split_at(prompt.len()).1.to_string());
    }

    fn update_history(&mut self, prompt: &str) -> Result<(), TuiError> {
        // Do not accept empty prompts in the history
        if prompt.trim().is_empty() {
            return Ok(());
        }
        let prompt = prompt.trim().to_string();
        if !self.history.contains(&prompt) {
            self.history.push(prompt.to_string());
            self.index_history = self.history.len() - 1;
            self.action_tx
                .as_ref()
                .unwrap()
                .send(Action::NewSearchPrompt(prompt))?;
        }
        Ok(())
    }

    fn search(&mut self) -> Result<(), TuiError> {
        let o = self.input.value().to_string();

        match ValidSearchQuery::from(o.as_str(), &self.filters_directory) {
            Ok(search_query) => {
                self.update_history(&o)?;
                self.action_tx
                    .clone()
                    .unwrap()
                    .send(Action::Search(search_query))?;
            }

            Err(e) => {
                if let Error::Search(SearchError::Parse(ee)) = &e {
                    self.remaining_input = Some(ee.to_string());
                }

                self.action_tx
                    .as_ref()
                    .unwrap()
                    .send(Action::Notification(Notification::new(
                        log::Level::Error,
                        e.to_string(),
                    )))?;
            }
        };
        Ok(())
    }

    #[allow(dead_code)]
    fn pretty_error_message(error: &nom::Err<nom::error::Error<&str>>) -> String {
        match error {
            nom::Err::Incomplete(_) => "Cannot parse query".to_string(),
            nom::Err::Error(s) => format!("unexpected token '{}'", s.input),
            nom::Err::Failure(s) => format!("unexpected token '{}'", s.input),
        }
    }
}

impl Component for SearchComponent {
    fn register_action_handler(&mut self, tx: UnboundedSender<Action>) {
        self.action_tx = Some(tx);
    }

    fn id(&self) -> ComponentName {
        ComponentName::Search
    }

    fn handle_key_events(&mut self, key: KeyEvent) -> Result<Option<Action>, TuiError> {
        self.remaining_input = None;
        match key.code {
            KeyCode::Right => {
                if self.input.value().len() == self.input.cursor() {
                    if let Some(a) = &self.autocomplete {
                        let input = self.input.value().to_string();
                        let autocompleted = format!("{input}{a}");
                        self.input = self
                            .input
                            .clone()
                            .with_cursor(autocompleted.len())
                            .with_value(autocompleted);
                        self.autocomplete = None;
                        self.compiler_worker.cancel();
                    }
                }
                self.input.handle_event(&Event::Key(key));
            }
            KeyCode::Up => self.autocomplete(KeyCode::Up),
            KeyCode::Down => self.autocomplete(KeyCode::Down),
            KeyCode::Enter => {
                self.entered = Some(Instant::now());
                self.search()?;
                self.autocomplete = None;
            }
            _ if !key.modifiers.contains(KeyModifiers::CONTROL) => {
                if let Some(e) = self.entered {
                    if e.elapsed() < Duration::from_millis(10) {
                        self.input.handle_event(&Event::Key(KeyEvent::new(
                            KeyCode::Char(' '),
                            KeyModifiers::NONE,
                        )));
                    }
                    self.entered = None;
                }
                self.input.handle_event(&Event::Key(key));
                self.parse_input();
                self.autocomplete(KeyCode::Backspace);
            }
            _ => {
                self.input.handle_event(&Event::Key(key));
            }
        };
        Ok(None)
    }

    fn draw(&mut self, f: &mut Frame<'_>, rect: Rect, state: &State) -> Result<(), TuiError> {
        let padding = 1;
        let input: &str = self.input.value();
        let mut line: Line = match &self.remaining_input {
            Some(e) => {
                let parts = input.split_at(input.len() - e.len());
                Line::from(vec![
                    Span::raw(parts.0),
                    Span::styled(
                        parts.1,
                        Style::default()
                            .fg(state.theme.orange)
                            .not_bold()
                            .underlined(),
                    ),
                ])
            }
            None => Line::from(vec![input.into()]),
        };

        if let Some(a) = &self.autocomplete {
            line.push_span(Span::styled(
                a,
                Style::default().fg(state.theme.autocomplete).not_bold(),
            ));
        }

        let position_cursor = Position {
            x: rect.x + 2 + (self.input.visual_cursor()) as u16,
            y: rect.y + 1,
        };

        // let rect = rect;
        //if rect.width < input.len() as u16 {
        //    let lines = input.len().div_ceil((rect.width - 15) as usize) as u16;
        //    info!("Lines {:?}", lines);
        //    rect.height =  3 + lines;
        //    rect.y = rect.y - lines;
        //    position_cursor = Position {
        //        x: rect.x + 1 + (self.input.visual_cursor()) as u16,
        //        y: rect.y + 1,
        //    }
        //}

        let block = Block::default()
            .borders(Borders::ALL)
            .border_type(BorderType::Rounded)
            .padding(Padding::left(1))
            .title(" Search ");
        let block = self.make_block_focused_with_state(state, block);

        let selected_style = Paragraph::new(line)
            .block(Block::new().padding(Padding::left(padding)))
            .wrap(Wrap { trim: false });

        let paragraph = selected_style.block(block);
        if state.is_focused(self.id()) {
            f.set_cursor_position(position_cursor);
        }
        //f.render_widget(Clear, rect);
        f.render_widget(paragraph, rect);
        Ok(())
    }

    fn shortcuts(&self) -> Vec<Shortcut> {
        vec![
            Shortcut::new("↑↓", "History"),
            Shortcut::new("ENTER", "Search"),
        ]
    }
}

#[cfg(test)]
use crate::assert_draw;

#[test]
fn test_draw() {
    let mut component = SearchComponent::default();
    assert_draw!(component, 60, 3)
}

#[test]
fn test_empty_history_underflow() {
    let component = SearchComponent::new("from begin", vec![], PathBuf::from("."));
    assert_eq!(component.index_history, 0);
}

//#[cfg(test)]
//use quickcheck_macros::quickcheck;
//
//#[cfg(test)]
//#[quickcheck]
//fn test_responsiveness(width: u16, height: u16) -> bool {
//    true
//    info!(
//        "Testing responsiveness with width: {}, height: {}",
//        width, height
//    );
//    use crate::Theme;
//    use app::configuration::GlobalConfig;
//    use log::info;
//    use ratatui::{Terminal, backend::TestBackend};
//    let mut component = SearchComponent::default();
//
//    let temp_dir = tempfile::tempdir().unwrap();
//    let temp_path = temp_dir.path().to_path_buf();
//
//    let state = State::new(
//        "test",
//        Theme::light(),
//        &GlobalConfig {
//            path: temp_path.clone().join("config.json"),
//            yozefu_directory: temp_path.join("config"),
//            logs: None,
//            default_url_template: "".to_string(),
//            initial_query: "".to_string(),
//            theme: "light".to_string(),
//            clusters: indexmap::IndexMap::default(),
//            default_kafka_config: indexmap::IndexMap::default(),
//            history: vec![],
//            show_shortcuts: true,
//            export_directory: std::path::PathBuf::from(""),
//        },
//    );
//
//    let mut terminal = Terminal::new(TestBackend::new(width, height)).unwrap();
//    terminal
//        .draw(|frame| component.draw(frame, frame.area(), &state).unwrap())
//        .is_ok()
//}
