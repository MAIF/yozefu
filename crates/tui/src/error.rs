use lib::Error;
use std::{
    fmt::{self, Display, Formatter},
    num::TryFromIntError,
};
use tokio::sync::mpsc::error::SendError;

/// Wrapper of around the `lib::Error` struct.
//#[derive(Debug)]
pub struct TuiError(Error);

impl<T> From<SendError<T>> for TuiError {
    fn from(e: SendError<T>) -> Self {
        TuiError(Error::Error(e.to_string()))
    }
}

impl Display for TuiError {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

use std::fmt::Debug;

impl Debug for TuiError {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl From<&str> for TuiError {
    fn from(e: &str) -> Self {
        TuiError(Error::Error(e.to_string()))
    }
}

impl From<serde_json::Error> for TuiError {
    fn from(e: serde_json::Error) -> Self {
        TuiError(Error::SerdeError(e))
    }
}

impl From<Error> for TuiError {
    fn from(e: Error) -> Self {
        TuiError(e)
    }
}

impl From<TuiError> for Error {
    fn from(val: TuiError) -> Self {
        val.0
    }
}

impl From<std::io::Error> for TuiError {
    fn from(e: std::io::Error) -> Self {
        TuiError(Error::IoError(e))
    }
}

impl From<TryFromIntError> for TuiError {
    fn from(e: TryFromIntError) -> Self {
        TuiError(Error::Error(e.to_string()))
    }
}
