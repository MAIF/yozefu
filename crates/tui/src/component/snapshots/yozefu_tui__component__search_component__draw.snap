---
source: crates/tui/src/component/search_component.rs
expression: terminal.backend()
---
"╭ Search ──────────────────────────────────────────────────╮"
"│                                                          │"
"╰──────────────────────────────────────────────────────────╯"
